src/worktree.c

```diff
@@ -365,12 +366,24 @@ int git_worktree_add(git_worktree **out, git_repository *repo,
 		if ((err = git_reference_dup(&ref, wtopts.ref)) < 0)
 			goto out;
 	} else {
-		if ((err = git_repository_head(&head, repo)) < 0)
-			goto out;
-		if ((err = git_commit_lookup(&commit, repo, &head->target.oid)) < 0)
-			goto out;
-		if ((err = git_branch_create(&ref, repo, name, commit, false)) < 0)
+		if ((err = git_buf_printf(&ref_buf, "refs/heads/%s", name)) < 0)
 			goto out;
+		if (!git_reference_lookup(&ref, repo, ref_buf.ptr)) {
+			if (git_branch_is_checked_out(ref)) {
+				git_error_set(GIT_ERROR_WORKTREE,
+					      "reference %s is already checked out",
+					      ref_buf.ptr);
+				err = -1;
+				goto out;
+			}
+		} else {
+			if ((err = git_repository_head(&head, repo)) < 0)
+				goto out;
+			if ((err = git_commit_lookup(&commit, repo, &head->target.oid)) < 0)
+				goto out;
+			if ((err = git_branch_create(&ref, repo, name, commit, false)) < 0)
+				goto out;
+		}
 	}
```

Comment:
Hm. I feel like you've got a point with the `git_branch_is_checked_out` call, we're not doing that e.g. if `wtopts.ref` is set. Let me propose the following flow:

```
if (wtopts.ref) {
    git_reference_dup(&ref, wtopts.ref);
} else if (git_branch_lookup(&ref, name) == 0) {
    /* Use the ref */
} else {
    git_branche_create(&ref);
}

if (git_branch_is_checked_out(ref)) {
    error out;
}
```

I think we ought to give the user a option to turn off re-use of branches, though. So maybe we should add a flag to `git_worktree_add_options`?

Reply:
I think we are already checking if `wopts.refs` is checkouted at the beginning of the `git_worktree_add` function with below code:
```
if (wtopts.ref) {
		if (!git_reference_is_branch(wtopts.ref)) {
			git_error_set(GIT_ERROR_WORKTREE, "reference is not a branch");
			err = -1;
			goto out;
		}

		if (git_branch_is_checked_out(wtopts.ref)) {
			git_error_set(GIT_ERROR_WORKTREE, "reference is already checked out");
			err = -1;
			goto out;
		}
}
```
So that should be fine.
Should we add a `force` option as in `git worktree` command ?

Reply:
Ah, fair enough. Didn't read into the code and only saw this single instance of `if (wtopts.ref)` in the diff. I'd prefer to have that check whether it's checked out in a single location as proposed by me to make it easier to extend, bu that's definitely not something you'll have to fix.

I'd like to not call this `force`, as this is a frequently overloaded name and doesn't really tell what's being forced. Something like "allow_reusing_refs" or "allow_existing_branch" would be better? I'm open for better proposals though!

Reply:
After thinking a bit more about it, the `force` flag would make sense only if the branch passed as `wtopts.ref` or inferred from `name` of the worktree is already checked out.
This pull request currently aborts if the branch is already checked out.
If the branch is not checked out, it will simply reuse it. So I guess that's fine from an user point of view.
Let me know what you think.

Reply:
> This pull request currently aborts if the branch is already checked out.

Let's keep it this way for now. Can still add it later if requested.